dnf/transaction_sr.py

```diff
@@ -0,0 +1,508 @@
+            return
+
+        # a cache to check no extra packages were pulled into the transaction
+        if action != "Reason Change":
+            self._nevra_cache.add(nevra)
+
+        # store reasons for forward actions and "Removed", the rest of the
+        # actions reasons should stay as they were determined by the transaction
+        if action in ("Install", "Upgrade", "Downgrade", "Reinstall", "Removed"):
+            self._nevra_reason_cache[nevra] = reason
+
+        if action in ("Install", "Upgrade", "Downgrade"):
+            if action == "Install" and query_na.installed() and not self._base._get_installonly_query(query_na):
+                self._raise_or_warn(self._ignore_installed,
+                    _('Rpm "{na}" is already installed for action "{action}".').format(na=na, action=action))
+
+            sltr = dnf.selector.Selector(self._base.sack).set(pkg=query)
+            self._base.goal.install(select=sltr, optional=self._base.conf.skip_broken)
+        elif action == "Reinstall":
+            query = query.available()
+
+            if not query:
+                self._raise_or_warn(self._skip_unavailable,
+                    _('Rpm nevra "{nevra}" not available in repositories for action "{action}".')
+                    .format(nevra=nevra, action=action))
+                return
+
+            sltr = dnf.selector.Selector(self._base.sack).set(pkg=query)
+            self._base.goal.install(select=sltr, optional=self._base.conf.skip_broken)
+        elif action in ("Upgraded", "Downgraded", "Reinstalled", "Removed", "Obsoleted"):
+            query = query.installed()
+
+            if not query:
+                # if ignore_installed is True, ignore the version of the package
+                # being removed and remove any NA
+                if action == "Removed" and self._ignore_installed:
+                    query_na = query_na.installed()
+
+                    if not query_na:
+                        self._warnings.append(
+                            _('Rpm "{na}" not installed for action "{action}".').format(na=na, action=action))
+                        return
+
+                    self._base.goal.erase(query_na[0], clean_deps=False)
+
+                    return
+
+                self._raise_or_warn(self._ignore_installed,
+                    _('Rpm nevra "{nevra}" not installed for action "{action}".').format(nevra=nevra, action=action))
+
+                return
+
+            # erasing the original version (the reverse part of an action like
+            # e.g. upgrade) is more robust, but we can't do it if
+            # skip_unavailable is True, because if the forward part of the
+            # action is skipped, we would simply remove the package here
+            if not self._skip_unavailable or action == "Removed":
+                self._base.goal.erase(query[0], clean_deps=False)
+        elif action == "Reason Change":
+            self._base.history.set_reason(query[0], reason)
```

Comment:
I don't know... what do you think?

Reply:
My question is what happen when package is not installed, but available:

1.  and it will be not installed in transaction?
2. but will be installed by the transaction?

Reply:
> My question is what happen when package is not installed, but available:
> 
>     1. and it will be not installed in transaction?

It creates a transaction that contains the reason change, even though the package isn't there...

>     2. but will be installed by the transaction?

This still ends up with two transaction items, one for the install and the other for the reason change...

3. reason change on a package that is also being removed in the transaction

Again, creates two transaction items...

I'm not sure what to do with this... How about prohibiting Reason Changes on packages that are not installed, as well prohibiting a Reason Change action on a NEVRA that already has another action in the stored transaction?

Reply:
> How about prohibiting Reason Changes on packages that are not installed

I don't want to say prohibiting, but skipping would be better

> as well prohibiting a Reason Change action on a NEVRA that already has another action in the stored transaction?

I suggest that this is not a good idea that could lead to a different result than the original transaction.